File: lambda-functions/cred-check.js
/**
 * cred-check — Daglig credential sanity-check på alla aktiva kunder.
 *
 * Körs via EventBridge varje morgon 06:00 CET. Testar varje kunds
 * WP-credentials live via REST API, rapporterar till BigQuery
 * customer_cred_status, och larmar om credentialer går sönder
 * (rotation, WP-uppgradering, plugin-konflikt osv).
 *
 * Tre outputs:
 *  1. BigQuery insert: customer_cred_status (historik)
 *  2. Om någon kund ändrat status (OK → FAIL): SNS-larm
 *  3. CloudWatch log: full rapport
 */
const https = require('https');
const { BigQuery } = require('@google-cloud/bigquery');
const { SSMClient, GetParameterCommand } = require('@aws-sdk/client-ssm');
const { SNSClient, PublishCommand } = require('@aws-sdk/client-sns');

const REGION = process.env.AWS_REGION || 'eu-north-1';
const ssm = new SSMClient({ region: REGION });
const sns = new SNSClient({ region: REGION });

const PROJECT_ID = 'searchboost-485810';
const DATASET = 'seo_data';
const SNS_TOPIC = process.env.SNS_TOPIC_ARN
  || 'arn:aws:sns:eu-north-1:176823989073:seo-optimizer-alerts';

const ACTIVE_CUSTOMERS = [
  'searchboost', 'ilmonte', 'mobelrondellen', 'tobler',
  'traficator', 'jelmtech', 'smalandskontorsmobler', 'humanpower',
];

async function getParam(name, withDecryption = false) {
  try {
    const r = await ssm.send(new GetParameterCommand({ Name: name, WithDecryption: withDecryption }));
    return r.Parameter.Value;
  } catch (e) {
    return null;
  }
}

function httpsGet(urlString, headers = {}, timeout = 15000) {
  return new Promise((resolve) => {
    const u = new URL(urlString);
    const req = https.request({
      hostname: u.hostname,
      port: 443,
      path: u.pathname + u.search,
      method: 'GET',
      headers: { 'User-Agent': 'Mozilla/5.0 SBS-CredCheck-Lambda', ...headers },
      timeout,
    }, (res) => {
      let body = '';
      res.on('data', (c) => (body += c));
      res.on('end', () => {
        try { resolve({ status: res.statusCode, json: JSON.parse(body) }); }
        catch { resolve({ status: res.statusCode, json: null, body }); }
      });
    });
    req.on('error', (e) => resolve({ status: null, error: e.message }));
    req.on('timeout', () => { req.destroy(); resolve({ status: null, error: 'timeout' }); });
    req.end();
  });
}

async function checkCustomer(cust) {
  const out = {
    customer_id: cust,
    checked_at: new Date().toISOString(),
    url: null, user: null,
    auth_ok: false, role: null,
    can_edit_posts: false,
    can_publish_posts: false,
    can_manage_options: false,
    code_snippets: false,
    rank_math: false,
    woocommerce: false,
    error: null,
  };

  const url = await getParam(`/seo-mcp/wordpress/${cust}/url`);
  const user = await getParam(`/seo-mcp/wordpress/${cust}/username`);
  const app = await getParam(`/seo-mcp/wordpress/${cust}/app-password`, true);

  out.url = url;
  out.user = user;

  if (!url || !user || !app || app === 'placeholder') {
    out.error = 'ssm_creds_missing';
    return out;
  }

  const auth = Buffer.from(`${user}:${app}`).toString('base64');
  const headers = { Authorization: `Basic ${auth}` };

  const me = await httpsGet(`${url.replace(/\/$/, '')}/wp-json/wp/v2/users/me?context=edit`, headers);
  if (me.status !== 200 || !me.json) {
    out.error = `auth_failed_${me.status || 'network'}`;
    return out;
  }

  out.auth_ok = true;
  const roles = me.json.roles || [];
  out.role = roles[0] || null;
  const caps = me.json.capabilities || {};
  out.can_edit_posts = !!caps.edit_posts;
  out.can_publish_posts = !!caps.publish_posts;
  out.can_manage_options = !!caps.manage_options;

  const ns = await httpsGet(`${url.replace(/\/$/, '')}/wp-json/`, headers);
  if (ns.status === 200 && ns.json) {
    const namespaces = ns.json.namespaces || [];
    out.code_snippets = namespaces.includes('code-snippets/v1');
    out.rank_math = namespaces.includes('rankmath/v1');
    out.woocommerce = namespaces.some((n) => n.startsWith('wc/'));
  }

  return out;
}

async function getPreviousStatus(bq) {
  try {
    const [rows] = await bq.query({
      query: `
        SELECT customer_id, auth_ok
        FROM \`${PROJECT_ID}.${DATASET}.customer_cred_status\`
        WHERE checked_at = (
          SELECT MAX(checked_at) FROM \`${PROJECT_ID}.${DATASET}.customer_cred_status\`
          WHERE checked_at < CURRENT_TIMESTAMP()
        )
      `,
    });
    const map = {};
    for (const r of rows) map[r.customer_id] = r.auth_ok;
    return map;
  } catch (err) {
    console.log('No previous status (first run?):', err.message);
    return {};
  }
}

async function alert(subject, message) {
  try {
    await sns.send(new PublishCommand({ TopicArn: SNS_TOPIC, Subject: subject.slice(0, 100), Message: message }));
  } catch (e) {
    console.error('SNS publish failed:', e.message);
  }
}

exports.handler = async () => {
  // BQ credentials
  const bqCredRaw = await getParam('/seo-mcp/bigquery/credentials', true);
  const bqCred = JSON.parse(bqCredRaw);
  const bq = new BigQuery({ projectId: PROJECT_ID, credentials: bqCred });

  // Ensure table exists
  const tableRef = `${PROJECT_ID}.${DATASET}.customer_cred_status`;
  await bq.query({
    query: `
      CREATE TABLE IF NOT EXISTS \`${tableRef}\` (
        checked_at TIMESTAMP,
        customer_id STRING,
        url STRING,
        user STRING,
        auth_ok BOOL,
        role STRING,
        can_edit_posts BOOL,
        can_publish_posts BOOL,
        can_manage_options BOOL,
        code_snippets BOOL,
        rank_math BOOL,
        woocommerce BOOL,
        error STRING
      )
      PARTITION BY DATE(checked_at)
      CLUSTER BY customer_id
    `,
  });

  const prev = await getPreviousStatus(bq);

  // Check all customers in parallel
  const results = await Promise.all(ACTIVE_CUSTOMERS.map(checkCustomer));

  // Insert to BQ
  await bq.dataset(DATASET).table('customer_cred_status').insert(results);

  // Detect status changes
  const newlyFailed = [];
  const recovered = [];
  for (const r of results) {
    const was = prev[r.customer_id];
    if (was === true && !r.auth_ok) newlyFailed.push(r);
    if (was === false && r.auth_ok) recovered.push(r);
  }

  if (newlyFailed.length > 0) {
    const msg = [
      'KRITISKT: WP-credentials slutade fungera på en eller flera kunder',
      '',
      ...newlyFailed.map((r) => `${r.customer_id}: ${r.error} (${r.url})`),
      '',
      'Kör cred_check.py lokalt för att verifiera + åtgärda i SSM.',
    ].join('\n');
    await alert(`[CRED-CHECK] ${newlyFailed.length} kunder tappade access`, msg);
  }

  if (recovered.length > 0) {
    const msg = recovered.map((r) => `${r.customer_id} (${r.url})`).join('\n');
    await alert(`[CRED-CHECK] ${recovered.length} kunder återfick access`, msg);
  }

  const ok = results.filter((r) => r.auth_ok).length;
  const fail = results.length - ok;
  console.log(`✓ Cred-check klar: ${ok} OK / ${fail} fail`);
  return {
    statusCode: 200,
    total: results.length,
    ok, fail,
    newly_failed: newlyFailed.map((r) => r.customer_id),
    recovered: recovered.map((r) => r.customer_id),
  };
};

File: scripts/cred_check.py
#!/usr/bin/env python3
"""
Credential sanity-checker för alla Searchboost-kunder.

Kör session-start eller innan rapportering av "creds saknas".
Testar varje kunds WP-credentials LIVE via REST och rapporterar
faktisk status: kan skriva / kan läsa / role / capabilities.

Resultat sparas i /tmp/sbs_cred_check_{date}.json så rapporten
kan läsas av Claude vid behov under sessionen.

Usage:
    python3 cred_check.py             # kör på alla aktiva kunder
    python3 cred_check.py ilmonte     # bara en kund
"""

import json
import sys
import base64
import subprocess
import urllib.request
import urllib.error
from datetime import datetime

ACTIVE_CUSTOMERS = [
    "searchboost",
    "ilmonte",
    "mobelrondellen",
    "tobler",
    "traficator",
    "jelmtech",
    "smalandskontorsmobler",
    "humanpower",
]

def ssm(name, dec=False):
    cmd = [
        "aws", "ssm", "get-parameter",
        "--name", name,
        "--region", "eu-north-1",
        "--profile", "mickedanne@gmail.com",
        "--query", "Parameter.Value",
        "--output", "text",
    ]
    if dec:
        cmd.append("--with-decryption")
    try:
        return subprocess.check_output(cmd, stderr=subprocess.DEVNULL).decode().strip()
    except subprocess.CalledProcessError:
        return None

def req(method, url, auth, timeout=15):
    r = urllib.request.Request(
        url, method=method,
        headers={
            "Authorization": f"Basic {auth}",
            "User-Agent": "Mozilla/5.0 (SBS-CredCheck)",
            "Accept": "application/json",
        },
    )
    try:
        with urllib.request.urlopen(r, timeout=timeout) as resp:
            return resp.status, json.loads(resp.read().decode() or "{}")
    except urllib.error.HTTPError as e:
        try:
            return e.code, json.loads(e.read().decode() or "{}")
        except Exception:
            return e.code, {}
    except Exception as e:
        return None, {"error": str(e)[:200]}

def check(cust):
    result = {
        "customer": cust,
        "url": None,
        "user": None,
        "password_present": False,
        "auth_ok": False,
        "role": None,
        "can_read": False,
        "can_edit_posts": False,
        "can_publish_posts": False,
        "can_manage_options": False,
        "can_edit_files": False,
        "code_snippets_installed": False,
        "rank_math_installed": False,
        "woocommerce_installed": False,
        "error": None,
    }

    url = ssm(f"/seo-mcp/wordpress/{cust}/url")
    user = ssm(f"/seo-mcp/wordpress/{cust}/username")
    app = ssm(f"/seo-mcp/wordpress/{cust}/app-password", dec=True)

    result["url"] = url
    result["user"] = user
    result["password_present"] = bool(app and app != "placeholder" and len(app) >= 10)

    if not (url and user and result["password_present"]):
        result["error"] = "SSM creds missing/placeholder"
        return result

    auth = base64.b64encode(f"{user}:{app}".encode()).decode()
    base = url.rstrip("/")

    # 1. whoami via wp/v2/users/me?context=edit
    code, me = req("GET", f"{base}/wp-json/wp/v2/users/me?context=edit", auth)
    if code != 200:
        result["error"] = f"auth failed [{code}] {me.get('message', me.get('code', ''))}"
        return result

    result["auth_ok"] = True
    result["can_read"] = True
    roles = me.get("roles") or []
    result["role"] = roles[0] if roles else None
    caps = me.get("capabilities") or {}
    result["can_edit_posts"] = bool(caps.get("edit_posts"))
    result["can_publish_posts"] = bool(caps.get("publish_posts"))
    result["can_manage_options"] = bool(caps.get("manage_options"))
    result["can_edit_files"] = bool(caps.get("edit_files"))

    # 2. Plugin check
    code, ns_root = req("GET", f"{base}/wp-json/", auth)
    if code == 200:
        namespaces = ns_root.get("namespaces") or []
        result["code_snippets_installed"] = "code-snippets/v1" in namespaces
        result["rank_math_installed"] = "rankmath/v1" in namespaces
        result["woocommerce_installed"] = any(n.startswith("wc/") for n in namespaces)

    return result

def main():
    targets = sys.argv[1:] if len(sys.argv) > 1 else ACTIVE_CUSTOMERS
    results = []
    for c in targets:
        print(f"\n=== {c} ===")
        r = check(c)
        results.append(r)
        if r["error"]:
            print(f"  ✗ {r['error']}")
        else:
            flags = []
            if r["can_edit_posts"]:
                flags.append("edit")
            if r["can_publish_posts"]:
                flags.append("publish")
            if r["can_manage_options"]:
                flags.append("manage")
            plugins = []
            if r["code_snippets_installed"]:
                plugins.append("code-snippets")
            if r["rank_math_installed"]:
                plugins.append("rank-math")
            if r["woocommerce_installed"]:
                plugins.append("woocommerce")
            print(f"  ✓ user={r['user']} role={r['role']} caps={','.join(flags)} plugins={','.join(plugins)}")

    outfile = f"/tmp/sbs_cred_check_{datetime.now().strftime('%Y%m%d')}.json"
    with open(outfile, "w") as f:
        json.dump({"checked_at": datetime.now().isoformat(), "results": results}, f, indent=2, ensure_ascii=False)
    print(f"\n→ resultat sparat: {outfile}")

if __name__ == "__main__":
    main()
